File: laboneq/dsl/calibration/amplifier_pump.py
# Copyright 2022 Zurich Instruments AG
# SPDX-License-Identifier: Apache-2.0

from __future__ import annotations

from dataclasses import dataclass, field

from laboneq.core.utilities.dsl_dataclass_decorator import classformatter
from laboneq.data.calibration import CancellationSource
from laboneq.dsl.calibration.observable import Observable
from laboneq.dsl.parameter import Parameter

amplifier_pump_id = 0


def amplifier_pump_id_generator():
    global amplifier_pump_id
    retval = f"ap{amplifier_pump_id}"
    amplifier_pump_id += 1
    return retval


@classformatter
@dataclass(init=True, repr=True, order=True)
class AmplifierPump(Observable):
    """Settings for the SHF Parametric Pump Controller (SHFPPC).

    Attributes:
        uid (str):
            Unique identifier. If left blank, a new unique ID will be
            generated.
        pump_frequency (float | Parameter | None):
            Set the pump frequency node. Default `None`.
        pump_power (float | Parameter | None):
            Set the pump power node. Units: dBm. Default `None`.
        pump_on (bool):
            Enable the pump tone. Default `True`.
        pump_filter_on (bool):
            Enable the integrated low-pass filter for the pump tone. Default: `True`.
        cancellation_on (bool):
            Enable pump tone cancellation. Default `True`.
        cancellation_phase (float | Parameter | None):
            Set the phase shift of the cancellation tone. Units: radians. Default `None`.
        cancellation_attenuation (float | Parameter | None):
            Set the attenuation of the cancellation tone. Positive values _reduce_ the
            cancellation tone power. Default `None`.
        cancellation_source (CancellationSource):
            Set the source of the cancellation tone. Default: internal.
        cancellation_source_frequency (float | None):
            Specify the cancellation tone frequency when using the *external*
            cancellation tone generator. Leave at `None` when supplying the
            cancellation tone internally (the frequency then matches that of the
            pump tone).
        alc_on (bool):
            Enable the automatic level control for pump tone output. Default `True`.
        probe_on (bool):
            Enable probe tone output. Default `False`.
        probe_frequency (float | Parameter | None):
            Set the frequency of the generated probe tone. Required if `probe_on` is
            `True`. Units: Hz. Default: `None`.
        probe_power (float | Parameter | None):
            Set the output power of the generated probe tone.
            Units: dBm. Default: `None`.

    Notes:
        If an attribute is set to `None`, the corresponding node is not set.

    !!! version-changed "Some fields were renamed in version 2.24.0"

        - `AmplifierPump.pump_freq` is now `AmplifierPump.pump_frequency`
        - `AmplifierPump.pump_engaged` is now `AmplifierPump.pump_on`
        - `AmplifierPump.alc_engaged` is now `AmplifierPump.alc_on`
        - `AmplifierPump.use_probe` is now `AmplifierPump.probe_on`
    """

    #: Unique identifier. If left blank, a new unique ID will be generated.
    uid: str = field(default_factory=amplifier_pump_id_generator)

    pump_frequency: float | Parameter | None = None
    pump_power: float | Parameter | None = None
    pump_on: bool = True
    pump_filter_on: bool = True
    cancellation_on: bool = True
    cancellation_phase: float | Parameter | None = None
    cancellation_attenuation: float | Parameter | None = None
    cancellation_source: CancellationSource = CancellationSource.INTERNAL
    cancellation_source_frequency: float | None = None
    alc_on: bool = True
    probe_on: bool = False
    probe_frequency: float | Parameter | None = None
    probe_power: float | Parameter | None = None

File: src/rust/laboneq-dsl/src/types/amplifier_pump.rs
// Copyright 2026 Zurich Instruments AG
// SPDX-License-Identifier: Apache-2.0

use std::str::FromStr;

use crate::types::{DeviceUid, ValueOrParameter};

#[derive(Debug, Clone, PartialEq)]
pub struct AmplifierPump {
    pub device: DeviceUid,
    pub channel: u16,
    pub alc_on: bool,
    pub pump_on: bool,
    pub pump_filter_on: bool,
    pub pump_power: Option<ValueOrParameter<f64>>,
    pub pump_frequency: Option<ValueOrParameter<f64>>,
    pub probe_on: bool,
    pub probe_power: Option<ValueOrParameter<f64>>,
    pub probe_frequency: Option<ValueOrParameter<f64>>,
    pub cancellation_on: bool,
    pub cancellation_phase: Option<ValueOrParameter<f64>>,
    pub cancellation_attenuation: Option<ValueOrParameter<f64>>,
    pub cancellation_source: PumpCancellationSource,
    pub cancellation_source_frequency: Option<f64>,
}

impl AmplifierPump {
    pub fn values_or_parameters(&self) -> impl Iterator<Item = &Option<ValueOrParameter<f64>>> {
        [
            &self.pump_power,
            &self.pump_frequency,
            &self.probe_power,
            &self.probe_frequency,
            &self.cancellation_phase,
            &self.cancellation_attenuation,
        ]
        .into_iter()
    }
}

#[derive(Debug, Clone, PartialEq, Copy, Eq, Default)]
pub enum PumpCancellationSource {
    #[default]
    Internal,
    External,
}

impl FromStr for PumpCancellationSource {
    type Err = String;

    fn from_str(s: &str) -> Result<Self, Self::Err> {
        match s.to_lowercase().as_str() {
            "internal" => Ok(PumpCancellationSource::Internal),
            "external" => Ok(PumpCancellationSource::External),
            _ => Err(format!("Unknown pump cancellation source: {}", s)),
        }
    }
}

pub struct AmplifierPumpBuilder {
    inner: AmplifierPump,
}

impl AmplifierPumpBuilder {
    pub fn new(device: DeviceUid, channel: u16) -> Self {
        Self {
            inner: AmplifierPump {
                device,
                channel,
                pump_frequency: None,
                pump_power: None,
                probe_power: None,
                probe_frequency: None,
                cancellation_phase: None,
                cancellation_attenuation: None,
                cancellation_source: PumpCancellationSource::default(),
                cancellation_source_frequency: None,
                alc_on: false,
                pump_on: false,
                pump_filter_on: false,
                probe_on: false,
                cancellation_on: false,
            },
        }
    }

    pub fn pump_frequency(mut self, value: ValueOrParameter<f64>) -> Self {
        self.inner.pump_frequency = Some(value);
        self
    }

    pub fn pump_power(mut self, value: ValueOrParameter<f64>) -> Self {
        self.inner.pump_power = Some(value);
        self
    }

    pub fn probe_frequency(mut self, value: ValueOrParameter<f64>) -> Self {
        self.inner.probe_frequency = Some(value);
        self
    }

    pub fn probe_power(mut self, value: ValueOrParameter<f64>) -> Self {
        self.inner.probe_power = Some(value);
        self
    }

    pub fn cancellation_phase(mut self, value: ValueOrParameter<f64>) -> Self {
        self.inner.cancellation_phase = Some(value);
        self
    }

    pub fn cancellation_attenuation(mut self, value: ValueOrParameter<f64>) -> Self {
        self.inner.cancellation_attenuation = Some(value);
        self
    }

    pub fn build(self) -> AmplifierPump {
        self.inner
    }
}
